Stacked bar chart

JohnUUU opened the pull request:
# Notable Code Changes 

1.  We split the bar-chart-series into 2 different dataseries: 
  - **Bar-chart-series** (Bar charts and Frequency charts)
  - **Multi-bar-chart-series** (Grouped Bar charts and Stacked Bar charts). 

2. In order to support the axis-format function without having to rely on google bar chart's ISU format string [Which sometimes breaks in a way that we can't catch on Pyret's side]. The way we did it was we analyzed the axis patterns for different bar lengths [here](https://docs.google.com/spreadsheets/d/1vCeQlOGQv-wXw05vbpl4Zryu7XzkDn-BLWrDUjokGaI/edit?usp=sharing), and found out that google bar chart looks at the (bar length / 9) to determine the gridline step length. We used that to calculate and generate the axis. 

3. We added some more edge case error checks for grouped-bar-charts and bar-charts and now raise an error message  
- When the data is empty, the labels and values are of different lengths for bar-chart-series 
- When the data is empty, the data groups have different lengths, labels and values have different lengths,  and legends and values have different lengths. 

4. We implemented a way to create stacked-bar-charts namely 
``` 
from-list.stacked-bar-chart(
      labels :: List<String>, 
      value-lists :: List<List<Number>>, 
      legends :: List<String>) 
```
![download (4)](https://user-images.githubusercontent.com/74892151/106087544-d9a44500-60e9-11eb-8aa8-c20f5b5d8421.png)

# Methods Implemented
We implemented the following methods. 
### Tags:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[single] - Supported for bar-chart-series 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[multi] - Supported for multi-bar-chart-series
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[issue] - Fails under some edge case or has some other issue

 
## Color Methods 
1. **Default Color** [single] - Changes the default color of all the bars in the series if another color is not specifically assigned to the bar. 
`bar-chart-series.default-color(c :: Color)`
<img width="675" alt="Screen Shot 2021-01-27 at 9 58 03 PM" src="https://user-images.githubusercontent.com/74892151/106088020-c6de4000-60ea-11eb-947f-319f1a144fe4.png">

2. **Colors** [single][multi][issue] - Specifies the specific colors of each bar or each bar component 
`bar-chart-series.colors(loc :: List<Color>)`
<img width="675" alt="Screen Shot 2021-01-27 at 10 02 24 PM" src="https://user-images.githubusercontent.com/74892151/106088318-600d5680-60eb-11eb-8ae8-dff71b71f09b.png">

`multi-bar-chart-series.colors(loc :: List<Color>)`
<img width="675" alt="Screen Shot 2021-01-27 at 10 04 58 PM" src="https://user-images.githubusercontent.com/74892151/106088542-c8f4ce80-60eb-11eb-8907-d817f4dd2c5d.png">
<img width="675" alt="Screen Shot 2021-01-27 at 10 06 44 PM" src="https://user-images.githubusercontent.com/74892151/106088657-fccff400-60eb-11eb-94d5-3dfdb5e88f35.png">

## Sorting Methods 
1. **Sort-by** [single][multi] - Sorts the bars using a comparator and equality checker by bar height 
`bar-chart-series.sort-by(cmp :: (Number, Number -> Boolean), eq :: (Number, Number -> Boolean))`
<img width="654" alt="Screen Shot 2021-01-27 at 10 15 34 PM" src="https://user-images.githubusercontent.com/74892151/106089372-76b4ad00-60ed-11eb-89b0-99d69af9cee6.png">

`multi-bar-chart-series.sort-by(cmp :: (Number, Number -> Boolean), eq :: (Number, Number -> Boolean))`
<img width="660" alt="Screen Shot 2021-01-27 at 10 32 01 PM" src="https://user-images.githubusercontent.com/74892151/106090315-7e755100-60ef-11eb-8d5f-f80722c2db68.png">
<img width="660" alt="Screen Shot 2021-01-27 at 10 30 00 PM" src="https://user-images.githubusercontent.com/74892151/106090183-36eec500-60ef-11eb-9420-7987e7ea387f.png">

2. **Sort-by-label** [single][multi] - Sorts the bars using a comparator and equality checker by bar name
`bar-chart-series.sort-by-label(cmp :: (String, String -> Boolean), eq :: (Sring, String -> Boolean))`
<img width="654" alt="Screen Shot 2021-01-27 at 10 22 12 PM" src="https://user-images.githubusercontent.com/74892151/106089702-1e31df80-60ee-11eb-960d-b88f7e31ad3a.png">

`multi-bar-chart-series.sort-by-label(cmp :: (String, String -> Boolean), eq :: (Sring, String -> Boolean))`
<img width="660" alt="Screen Shot 2021-01-27 at 10 35 30 PM" src="https://user-images.githubusercontent.com/74892151/106090531-fba0c600-60ef-11eb-8928-e3a59a8b47c9.png">
<img width="660" alt="Screen Shot 2021-01-27 at 10 36 12 PM" src="https://user-images.githubusercontent.com/74892151/106090552-1410e080-60f0-11eb-9726-63b02a8f6987.png">

3. **Sort-by-data** [multi] - Scores the bars using a scorer that looks at all of the bar components. Then sorts the bars by score using a comparator and equality checker. Example is sorting by first bar component. 
```
multi-bar-chart-series.sort-by-data<A>(
      scorer :: (List<Number> -> A), 
      cmp :: (A, A -> Boolean), 
      eq :: (A, A -> Boolean))
```
<img width="669" alt="Screen Shot 2021-01-27 at 10 40 56 PM" src="https://user-images.githubusercontent.com/74892151/106090843-bcbf4000-60f0-11eb-96ee-5427780a7fe8.png">
<img width="669" alt="Screen Shot 2021-01-27 at 10 41 39 PM" src="https://user-images.githubusercontent.com/74892151/106090881-d791b480-60f0-11eb-8a6c-f2d3a31d2a13.png">

## Axis Methods
1. **Add Pointers**[single][multi][issue] - Adds pointers at certain ticks on the axis to highlight certain lines of interest 
`bar-chart-series.add-pointers(ticks :: List<Number>, labels :: List<String>)`
<img width="679" alt="Screen Shot 2021-01-27 at 10 52 04 PM" src="https://user-images.githubusercontent.com/74892151/106091546-4b808c80-60f2-11eb-9344-3f7318a74eda.png">

`multi-bar-chart-series.add-pointers(ticks :: List<Number>, labels :: List<String>)`
<img width="728" alt="Screen Shot 2021-01-27 at 10 53 36 PM" src="https://user-images.githubusercontent.com/74892151/106091684-897db080-60f2-11eb-9c92-6d9608a8df16.png">
<img width="699" alt="Screen Shot 2021-01-27 at 10 54 30 PM" src="https://user-images.githubusercontent.com/74892151/106091741-a31ef800-60f2-11eb-8ea0-89edbc2d16a2.png">

2. **Pointer Color**[single][multi] - Colors the pointers added by Add Pointers
`bar-chart-series.pointer-color(c :: Color)`
<img width="699" alt="Screen Shot 2021-01-27 at 11 08 02 PM" src="https://user-images.githubusercontent.com/74892151/106096337-f7c67100-60fa-11eb-86cf-ad0df71cffa0.png">

`multi-bar-chart-series.pointer-color(c :: Color)`
<img width="806" alt="Screen Shot 2021-01-27 at 11 10 25 PM" src="https://user-images.githubusercontent.com/74892151/106092893-dc586780-60f4-11eb-93c2-819d46cf800d.png">
<img width="741" alt="Screen Shot 2021-01-27 at 11 11 14 PM" src="https://user-images.githubusercontent.com/74892151/106092956-fa25cc80-60f4-11eb-892a-0d4e0a686232.png">

3. **Format Axis**[single][multi] - Changes the names of the ticks on the axis 
`bar-chart-series.format-axis(f :: (Number -> String))`
<img width="680" alt="Screen Shot 2021-01-27 at 11 16 39 PM" src="https://user-images.githubusercontent.com/74892151/106093352-baabb000-60f5-11eb-890f-573ee8aead98.png">

`multi-bar-chart-series.format-axis(f :: (Number -> String))`
<img width="695" alt="Screen Shot 2021-01-27 at 11 17 48 PM" src="https://user-images.githubusercontent.com/74892151/106093426-e29b1380-60f5-11eb-88ea-f5db85a0b51b.png">
<img width="695" alt="Screen Shot 2021-01-27 at 11 18 54 PM" src="https://user-images.githubusercontent.com/74892151/106093509-0b230d80-60f6-11eb-866c-d3b89fe111a1.png">

## Scale Method 
1. **scale**[single][multi] - Scales the data by the function. Example is taking the num-log of the data. 
`bar-chart-series.scale(f :: (Number -> Number))`
<img width="676" alt="Screen Shot 2021-01-27 at 11 22 07 PM" src="https://user-images.githubusercontent.com/74892151/106093761-7ec51a80-60f6-11eb-812d-78c41372a834.png">

`multi-bar-chart-series.scale(f :: (Number -> Number))`
<img width="676" alt="Screen Shot 2021-01-27 at 11 23 40 PM" src="https://user-images.githubusercontent.com/74892151/106093870-b59b3080-60f6-11eb-9b12-3bf95b26c0d6.png">
<img width="637" alt="Screen Shot 2021-01-27 at 11 25 03 PM" src="https://user-images.githubusercontent.com/74892151/106093998-e8452900-60f6-11eb-97d4-b2e53a3bdc0f.png">

## Bar Type Methods
1. **Stacking-type**[multi] - Switches between stacking types. Stacking types are shown below. 
`multi-bar-chart-series.stacking-type(type :: String)`
<img width="608" alt="Screen Shot 2021-01-28 at 12 40 35 AM" src="https://user-images.githubusercontent.com/74892151/106100243-9fdf3880-6101-11eb-9e71-d28d18c93fd8.png">

2. **Horizontal**[single][multi] - Switches between bar and column chart. Options are shown below. 
`bar-chart-series.horizontal(type :: Boolean)`
<img width="601" alt="Screen Shot 2021-01-28 at 12 44 50 AM" src="https://user-images.githubusercontent.com/74892151/106100490-0e23fb00-6102-11eb-9e90-a66901c833b2.png">

`multi-bar-chart-series.horizontal(type :: Boolean)`
<img width="629" alt="Screen Shot 2021-01-28 at 12 46 23 AM" src="https://user-images.githubusercontent.com/74892151/106100620-44fa1100-6102-11eb-87d3-58fa1894e825.png">
<img width="671" alt="Screen Shot 2021-01-28 at 12 48 19 AM" src="https://user-images.githubusercontent.com/74892151/106100793-85598f00-6102-11eb-90e1-f726ff7a323c.png">

## Descriptor Object Methods 

1. **Annotations**[single][multi][issue] - Adds annotations that describe certain bars and bar-components 
`bar-chart-series.annotations(ann :: List<Option<String>>)`
<img width="658" alt="Screen Shot 2021-01-27 at 11 35 08 PM" src="https://user-images.githubusercontent.com/74892151/106094827-51796c00-60f8-11eb-9654-3e5d5c97c2b4.png">

`multi-bar-chart-series.annotations(ann :: List<List<Option<String>>>)`
<img width="627" alt="Screen Shot 2021-01-27 at 11 46 35 PM" src="https://user-images.githubusercontent.com/74892151/106095771-f9dc0000-60f9-11eb-94dd-ac520995c9fa.png">
<img width="651" alt="Screen Shot 2021-01-27 at 11 49 29 PM" src="https://user-images.githubusercontent.com/74892151/106095944-4fb0a800-60fa-11eb-94e9-d54b7c051770.png">

2. **Interval**[single][multi][issue] - Adds an interval at the top of a bar or bar component 
`bar-chart-series.intervals(int :: List<List<Number>>)`
<img width="651" alt="Screen Shot 2021-01-28 at 12 05 01 AM" src="https://user-images.githubusercontent.com/74892151/106097214-7ff94600-60fc-11eb-8a32-4f47d4791456.png">

`multi-bar-chart-series.intervals(int :: List<List<List<Number>>>)`
<img width="609" alt="Screen Shot 2021-01-28 at 12 09 53 AM" src="https://user-images.githubusercontent.com/74892151/106097637-4aa12800-60fd-11eb-9963-0c3ce5c0e28b.png">
<img width="638" alt="Screen Shot 2021-01-28 at 12 10 48 AM" src="https://user-images.githubusercontent.com/74892151/106097643-4c6aeb80-60fd-11eb-913c-893e0c525e86.png">

3. **Error Bars**[single][multi][issue] - Adds an error-bar at the to of a bar or bar component
`bar-chart-series.error-bars(int :: List<List<Number>>)`
<img width="638" alt="Screen Shot 2021-01-28 at 12 16 34 AM" src="https://user-images.githubusercontent.com/74892151/106098108-18dc9100-60fe-11eb-9c9f-7394e6b9da66.png">

`multi-bar-chart-series.error-bars(int :: List<List<List<Number>>>)`
<img width="613" alt="Screen Shot 2021-01-28 at 12 17 00 AM" src="https://user-images.githubusercontent.com/74892151/106098149-2b56ca80-60fe-11eb-81db-539b4cbe250b.png">
<img width="638" alt="Screen Shot 2021-01-28 at 12 17 22 AM" src="https://user-images.githubusercontent.com/74892151/106098391-93a5ac00-60fe-11eb-8302-ed074f7a241d.png">

4. **Interval-color**[single][multi] - Colors the intervals 
`bar-chart-series.interval-color(c :: Color)`
<img width="642" alt="Screen Shot 2021-01-28 at 12 22 46 AM" src="https://user-images.githubusercontent.com/74892151/106098634-fc8d2400-60fe-11eb-9b03-ea0141ca1f9a.png">

`multi-bar-chart-series.interval-color(c :: Color)`
<img width="673" alt="Screen Shot 2021-01-28 at 12 22 52 AM" src="https://user-images.githubusercontent.com/74892151/106098648-0151d800-60ff-11eb-843a-3901a2581e92.png">
<img width="638" alt="Screen Shot 2021-01-28 at 12 21 26 AM" src="https://user-images.githubusercontent.com/74892151/106098660-044cc880-60ff-11eb-858a-6d8b9c825975.png">

# Issues 
1. The current **colors method** doesn't support coloring bars in the middle or end bars without touching the beginning bars 
2. Multibars are not very compatible with annotations in some situations because it's prone to being clustered, overlapping, and stacked-bar-charts alternate drawing bar-components and annotations so a lot of the annotations get overlayed by a bar component and is unreadable. Same thing happens with intervals and error-bars for stacked-bar-charts
<img width="882" alt="Screen Shot 2021-01-28 at 12 27 04 AM" src="https://user-images.githubusercontent.com/74892151/106098982-9523a400-60ff-11eb-8ff5-2caaefef7fba.png">

3. Add-Pointers: Sometimes the pointers will be to close to one another and one of them will fail to be displayed
<img width="350" alt="Screen Shot 2021-01-28 at 12 29 39 AM" src="https://user-images.githubusercontent.com/74892151/106099420-40ccf400-6100-11eb-97ae-d9479d190281.png">

4. Add-Pointers + Horizontal: The pointer names don't show up when the bar-chart is horizontal.
<img width="338" alt="Screen Shot 2021-01-28 at 12 33 03 AM" src="https://user-images.githubusercontent.com/74892151/106099701-ba64e200-6100-11eb-93a7-6a5a786682ed.png">

5. Travis CI takes around 50 seconds to run the tests for chart.arr. The other chart test suites take around 5 seconds to run, so we might want to reduce the tests to the basics and not test edge cases.

# CI Error 
There's an inconsistent CI timeout error that we can't seem to track down but we know that our implementations have passed previous CI tests and that it consistently makes it past the chart-test file, At this point though, we might as well just push what we have first.

JohnUUU committed:
Adding amaranth to color.arr

This commit is mainly to make sure that changes to our forked code.pyret.org repository actually show up in our remote version of code.pyret.org (Heroku).

JohnUUU committed:
Added edge case error checks to from-list.grouped-bar-chart

Updated the function to have a documentation string, a few comments, and fleshed out the edge case error checks to catch a few cases that it missed before.

eyanje committed:
Add stacked bar charts

Added a field is-stacked to BarChartSeries and a function
from-list.stacked-bar-chart which acts just like
from-list.grouped-bar-chart, but sets is-stacked to be true. The
is-stacked field is passed directly into Google's API.

JohnUUU committed:
Added color method to BarChartSeries

Added a color method to BarChartSeries that takes in a list of colors to
1. color the bars of each group in Grouped-bar-charts
2. color the stacks of each bar in the Stacked-bar-charts

Currently Messes things up for the regular bar-charts function though :(

eyanje committed:
Add s to colors

JohnUUU committed:
Seperated bar-chart-series into 2 seperate dataseries

**Current Update**:
Seperated bar-chart-series into
1) **Bar-chart-series**: [bar-chart, freq-bar-chart]
2) **Multi-Bar-Chart-Series**: [Grouped-bar-chart, Stacked-bar-chart]

Also sets up a **colors** method for both dataseries, using a list of colors to
1) color each bar in the Bar-chart-series
2) color each stack or group member in the Multi-bar-chart-series

**Suggestion For Future Development**:
Right now the  **colors** method
1) colors the bar from *Left to Right* in Bar-chart-series and Grouped-bar-charts
2) colors the bar from *Bottom to Top* in Stacked-bar-charts

and sticking to default colors for the rest if unspecified. This doesn't account for the case if users want to color a bar in the middle or end and want the begginning to be the default color, so for version 1.1 that's something we might want to consider.

JohnUUU committed:
Merge branch 'stacked-bar-chart-color' of https://github.com/JohnUUU/code.pyret.org into stacked-bar-chart-color

JohnUUU committed:
Added a default-color method for Bar-Chart-Series

The default-color method allows users to change the default color of their bars from the google-default.

Cases where this is Useful:
1) They can use this to quickly color all the bars to the color that they want
2) When used with the colors method they dont have to manually color all unspecified bars

Also I added some comments to seperate the different components of the bar chart functions (Mainly to help me quickly identify where to edit). If this becomes too much baggage at any time during the development process we can just remove it.

JohnUUU committed:
Merge pull request #1 from JohnUUU/stacked-bar-chart-color

Separated bar-chart-series into 2 dataseries -> multi-bar-chart-series and bar-chart-series 
Added color methods to both dataseries

eyanje committed:
Add chaart testing (and fix color entry)

eyanje committed:
Add stacked-bar-chart colors test

JohnUUU committed:
Added Section Comments to bar-chart-test.arr

JohnUUU committed:
Default Colors [Marker Colors]

This is setting up the Defaults to the teacher marker colors that Emmanuel recommended.

Personally, I find this color palette to be a little lacking when compared to google's defaults since the [Red, Green, and Blue] are too bright and the shift to Black and Brown seems too sudden.

We can always change/update the color palette later to be a little bit more refined while still prioritizing colors that teachers have in their marker collection.

JohnUUU committed:
Sorting Function Implementation

Implemented the following table sorting methods for each bar-series

BAR-CHART-SERIES:
sort-by: Uses the inputted cmp and eq to sort the bars by their height
sort-by-label: Uses the inputted cmp and eq to sort the bars by their labels

MULTI-BAR-CHART-SERIES:
sort-by: Uses the inputted cmp and eq to sort the bars by their height
sort-by-label: Uses the inputted cmp and eq to sort the bars by their labels
sort-by-data: Uses a scoring function to score the stacks of data then uses the inputted cmp and eq to sort the bars by their score.

Will add tests in a future commit

JohnUUU committed:
Flushed out Testing File

Split the tests for Single Bar Charts, Grouped Bar Charts, and Stacked Bar Charts into 3 different checks

Added more basic rendering tests from the google docs to this file. Plus added the sorting method tests

JohnUUU committed:
Fixed some tests typos

JohnUUU committed:
Format Method [UNSTABLE]

Implemented the Format Method by passing the format string along to google's API but there are some deformed strings that fail for google's API. Not sure what they are so I cant catch it on Pyret's side

JohnUUU committed:
Revert "Format Method [UNSTABLE]"

This reverts commit 12c51fe09e0254153cd71dcfa3070499d41d5b52.

JohnUUU committed:
Revert "Revert "Format Method [UNSTABLE]""

This reverts commit c1eeb05ed2a16cc0d461baa30ada617d701fd105.

JohnUUU committed:
Revert "Revert "Revert "Format Method [UNSTABLE]"""

This reverts commit 0e142d71521bb0d66d123e3f60cc5287f150ce9f.

eyanje committed:
Add annotations

eyanje committed:
Add annotations

eyanje committed:
Fix support for single bar chart annotations

Fixed a an earlier functions/method conflict so that now
annotations-method is called properly

eyanje committed:
Fix support for single bar chart annotations

Fixed a an earlier functions/method conflict so that now
annotations-method is called properly

eyanje committed:
Fix method call for single annotations

eyanje committed:
Fix method call for single annotations

JohnUUU committed:
Add-Pointer Method [Stacked-Bar-Chart] [Unfinished]

Implemented a Tentaive working add-pointer method that takes in a list of tick names, and a list of tick positions and creates "pointers" on a second axis for Stacked Bar Chart.

PROBLEM: When the tick positions are too close together one or both tick labels disappear.

UNFINISHED: I still have to update this to work with grouped bar charts and regular bar charts as well as import the tests from the google doc to the actual testing file. I'll get that done sometime tommorow

JohnUUU committed:
Finished Pointer Method [Kinda]

Sorry for the delay! Turns out the old pointer method was a little bit flawed because both axis either didn't match up or matched up really weirdly. So I used the pattern in 2-bar maximums to calculate a better axis top and bottom.

Still doesn't match google's behavior sometimes though so enforced the max and min of both axis using this method. Looks a lot more natural than the last commit though!

Other Notable Changes:
  - Added Egde Case Error Checkers
  - Implmented working versions for all bar-chart types
 -  Further fleshed out the test cases for all Bar Chart methods

Some graphs right now look a little bit off because some of the minor gridlines were excluded for some reason. However since we're calculating the stepsize and min/max of the axis anyways, it's probably smarter to do this in pyret so that we can write an axis-formatter that bypasses the ICU string thing. I'll include the minorgridlines there.

Plus will also work on creating an axis pointer coloring method!

eyanje committed:
Add intervals

JohnUUU committed:
Merge pull request #2 from JohnUUU/sbc-testing

Color Fix + Basic Testing Framework

eyanje committed:
Add error-bars methods

eyanje committed:
Add error-bars methods

eyanje committed:
Fix merge error

eyanje committed:
Specify dimension in error message

eyanje committed:
Specify dimension in error message

eyanje committed:
Set internal interval value

eyanje committed:
Merge with origin

eyanje committed:
Add intervals field

eyanje committed:
Add intervals fields

eyanje committed:
Merge with origin

eyanje committed:
Remove duplicate definitions

eyanje committed:
Rudimentary tests (no errors)

JohnUUU committed:
Finalized Axis Functions

1) Moved all the axis calculating functions into Pyret and used that to implment a non ICU format string format method
2) Added a Pointer Color method to color the pointers
3) Created more tests for format and pointer methods

JohnUUU committed:
Fixed Testing File

There was a unclosed parenthesis which was causing the build to fail

JohnUUU committed:
Dubious Scale Method

1) Updated axis creation to account for another edgecase
2) Implemented the Scale Method for most cases. The problem is that in order to do this I use the top of the axis and bottom of the axis and mutate them with the scale function as well. This leads to some cases where the data is in the domain of the scale function but the axis top and bottoms are not causing an error [SEE Num-log test case].
I'm not sure how to rectify this, I could have the method simply recreate the entire chart using the from-list method but the problem with that is that it would reset every other method as well.  Not sure where to go with this method from here, will raise an issue and implement the rest of the functions

JohnUUU committed:
Fixed Scale Function, axis calculation error, and added support for roughnums by extimating them into rational numbers

1) Rewrote the scale functions axis calculator to be dependent on the data and type rather than the previous axis [As it should have always been]
2) Fixed a bug in axis calculation where some numbers become BigIntegers and thus the chart library doesn't read it correctly.
3) Added support for charts with rational numbers and fleshed out tests for all methods with these new charts

JohnUUU committed:
Implemented stacking types

1) Implemented stacking types giving us access to relative and percent stacking + easy way to switch between grouped and stacked bars
2) Refactored and edited axis calculation to support each type of bar chart

JohnUUU committed:
Implemented Horizontal

Implemented the option to flip the axes of the chart aka the orientation. Works with basically all methods except add-pointers because of this isssue.
https://stackoverflow.com/questions/32059511/google-charts-dual-x-axis-different-ticks?answertab=active#tab-top
Can't use Material Bar charts too because that would mess up even more methods.

JohnUUU committed:
Merge pull request #6 from JohnUUU/sbc-orientation

eyanje committed:
Merge stacked-bar-chart into sbc-error-bars

eyanje committed:
Merge sbc-error-bars into stacked-bar-chart

Add support for annotations, intervals, and error bars

JohnUUU committed:
Fixed add-pointer, and intervals/error bars + custom axis, and added interval color method

Turns out the new annotations were throwing off the add-pointer method because they were adding a lot of extra role columns so I just set the target of all those columns to the second axis.
I updated the intervals method to work with the custom axis method so that the intervals dont get suddenly cut off from the top and bottom of the screen
Added a interval color method that lets you change the color of the intervals and make it stand out more.

sorawee commented on src/web/arr/trove/chart.arr:
```diff
@@ -122,6 +285,367 @@ y-max-method = method(self, y-max :: Number):
+  scaled-self.make-axis(max-positive-height, max-negative-height)
+end
+
+stacking-type-method = method(self, stack-type :: String): 
```
Does it make more sense to define a new data definition, rather than to use a string, for `stack-type`?

JohnUUU replied:
I think it makes some sense. 

The main difference between the two series is that 
- Bar-chart-series has its data represented by a List<Number> which contains the heights of the bars. 
- Multi-bar-chart-series has its data represented by a List<List<Number>> which contains the heights of the bars separated into components. 

The reason this is important is that many of the methods shown above must be implemented differently in each case. For example: 
- Colors is defined to color each bar in the bar-chart-series case and each component in the multi-bar-chart-series 
- Sort-by-data only makes sense in the context of a multi-bar-chart-series, where you might want to only sort by certain components 
- Calculating the axis depends on the highest bar displayed and the method needed to find the changes for each data-series. 
- Scale also has to be mapped differently based on each data-series 

Also stacking type would have to be able to change between the List<List<Number>> (multi-bar-chart-series) data and List<Number> (bar-chart-series) data. 
While we can switch from multi-bar-chart-series to bar-chart-series naturally,  It doesn’t seem feasible to go from bar-chart-series to multi-bar-chart-series naturally.

sorawee replied:
Hmm. I'm not sure if I understand your comment. What I meant is that we could define:

```
data StackType:
  | absolute
  | relative
  | percent
  | none
end
```

and use these instead of string.

JohnUUU replied:
Oh I'm sorry I misunderstood what you saying. I think that forcing the users to create an new stacktype instance every time they want to change stacking types might be a bit too cumbersome and maybe too complicated/uncommon for anyone just starting out learning pyret. Thus personally I believe using strings would be better if only because they would be more familiar.

sorawee replied:
> I think that forcing the users to create an new stacktype instance every time they want to change stacking types might be a bit too cumbersome

I don't think it's any more cumbersome than creating stuff like `mode-solid` in the `image-typed` library.

> maybe too complicated/uncommon for anyone just starting out learning pyret

My goal is to provide a good interface for people to use. If some audiences need a very simple interface, wrapper libraries could be created for that purpose, and this is already the case for the chart library for Bootstrap.

`Option` looks pretty complicated compared to `null`. One can argue that people learning programming would understand `null` better, and that might even be true. Still, there is value in learning the right way to do things.

JohnUUU replied:
Ah I see that makes sense, it's actually not cumbersome at all. Sorry, I must've gotten a bit confused from jumping around different coding languages, for some reason I thought you would have to make a new stacking type instance every time you called it like with classes. I actually have the opposite opinion now in that the strings might be unnecessary. Or should we do it the same was as Fillmode where we support both. If so should we split it into an typed and untyped library?

sorawee replied:
I think the fact that `Fillmode` is different in the typed and untyped library is mostly because the typed one is better and is what we want, but we can't remove the untyped one due to backward incompatibility. (the untyped interface has been around for a very long time)

One of the reasons why I am pushing to have `StackType` now is that if we wait longer and people start to use the current interface, then we won't be able to change the interface anymore due to the backward incompatibility constraint.

blerner replied:
It's not quite that simple, either. In principle I agree with you, but in practice, using enumerated values doesn't necessarily make things simpler for teaching. Now you have a curricular dependency on explaining what these magic variable names mean, whereas the string-based API only requires teaching about strings.

I think possibly the right thing to do here is make *this* API be an enumeration, and ask @schanzer to wrap the API in a stringier version for use in bootstrap teachpack libraries. @schanzer and @jpolitz , agreed?

sorawee commented on src/web/arr/trove/color.arr:
```diff
@@ -52,6 +52,7 @@ gold = color(255, 215, 0, 1)
 yellow = color(255, 255, 0, 1)
 olive = color(128, 128, 0, 1)
 burlywood = color(222, 184, 135, 1)
+amaranth = color(229, 43, 80, 1)
```
Just to give you a headsup, this also needs an update at the `pyret-docs` repo, specifically on this file: https://github.com/brownplt/pyret-docs/blob/master/src/trove/color.js.rkt

blerner replied:
Where did this color come from?  It's not an X11 color, or a DrRacket color, and I would like to have some sort of systematic reason for adding new colors...

JohnUUU replied:
We mainly added this color because we wanted to test if we set up our gitHub branch and local pyret repl correctly. I guess after we confirmed that our commits actually result in changes to the local repl, we just moved on and left that in. 

The reason that we chose Amaranth is that we to quickly find a color that wasn’t already on the list of colors so we looked up the most underrated colors and picked one we liked from [here](https://www.mentalfloss.com/article/53121/11-colors-youve-probably-never-heard)

I’m perfectly okay with removing that color now if it’s unnecessary and goes against the current system of colors.

sorawee commented on src/web/js/trove/chart-lib.js:
```diff
@@ -70,6 +71,28 @@
+  }
+
+  // Converts a list with func:
+  function convertListWith(func, l) {
```
I think this is unsafe. JS Stack might overflow. I'm also not sure if using `val.dict.first` as a terminating condition is actually OK. In any case, I think there already exists a similar but safe function in core Pyret already, so you should just use it. Another possibility is to convert a list to a (raw-)array from Pyret side.

JohnUUU replied:
Got it! I’ll convert the lists into raw-arrays on Pyret's side in the next commit.

sorawee commented on src/web/js/trove/chart-lib.js:
```diff
@@ -269,22 +396,251 @@
+    addAnnotations(data, rawData);
+    addIntervals(data, rawData);
+
+    var options = {
```
Use `let` or `const` if possible. Also, there's a missing semicolon.

JohnUUU replied:
Got it! Will fix it in the next commit!

JohnUUU committed:
Replaced Unsafe ConvertListWith function + Worked in Const, Let > Var

1. Converted the lists to arrays from Pyret's side rather than Javascript's side.

2. Moved out the Bar chart getter functions so that I could make as many variables into consts as I could. Made the other vars into lets.

sorawee commented on src/web/js/trove/chart-lib.js:
```diff
@@ -70,6 +71,11 @@
   }
 
+  function convertPointer(p) {
+    return {v: toFixnum(p.dict.value) , f: p.dict.label}
```
If I remember correctly, you shouldn't refer to `.dict` directly because it's an internal structure of Pyret objects that could potentially change in the future. Instead, use `get`.

JohnUUU replied:
Sorry for the late reply, I was a little bit busy the last few days. I'll quickly swap out the .dict with get in the next commit!

JohnUUU committed:
Replaced instances of .dict with get

Instead of calling .dict directly which could change in the future, used the getter function to get the fields instead

blerner commented on src/web/arr/trove/chart.arr:
````diff
@@ -75,13 +86,150 @@ fun to-table3(xs :: List<Any>, ys :: List<Any>, zs :: List<Any>) -> TableIntern:
+    raise("Num-to-scientific: Only defined on bases > 1")
+  end
+  
+  fun recur(s :: SciNumber): 
+    doc: ``` 
+         Takes the current Coeff, Exponent and divides/multiplies by base to move closer to 
+         the actual scientific representation.
+         ```
+    cases (SciNumber) s: 
+      | sci-notation(c, e, b) => 
+        pos-c = num-abs(c)
+        ask: 
+          | (pos-c > 0) and (pos-c < 1) then: recur(sci-notation(c * b, e - 1, b))
+          | (pos-c == 0) or ((pos-c >= 1) and (pos-c < b)) then: sci-notation(c, e, b)
+          | otherwise: recur(sci-notation(c / b, e + 1, b))
+        end
+    end
+  end
+  
+  {(n): recur(sci-notation(n, 0, base))}
````
Replace these lines with the more direct
```
  fun convert(n :: Number):
    if num-is-rational(n) and (n == 0): sci-notation(0, 0, base)
    else:
      pow = num-floor(num-log(num-abs(n)) / num-log(base))
      c = n / num-expt(base, pow)
      sci-notation(c, pow, base)
    end
  end
  convert
```

JohnUUU replied:
Got it, replacing the function now!

blerner commented on src/web/arr/trove/chart.arr:
```diff
@@ -75,13 +86,150 @@ fun to-table3(xs :: List<Any>, ys :: List<Any>, zs :: List<Any>) -> TableIntern:
+  ask:
+    | is-stacked == 'none' then: 
+      # Find the tallest bar in the entire group 
+      positive-max-groups = map({(l): get-with-cmp({(a, b): a > b}, l)}, value-lists)
+      negative-max-groups = map({(l): get-with-cmp({(a, b): a < b}, l)}, value-lists)
+      max-positive-height = num-max(0, get-with-cmp({(a, b): a > b}, positive-max-groups))
+      max-negative-height = num-min(0, get-with-cmp({(a, b): a < b}, negative-max-groups))
```
replace these lambdas with `num-max` and `num-min`, and then eliminate `get-with-cmp` in favor of calling `fold` directly

JohnUUU replied:
Good point, I guess this was an attempt at generalization that just ended up being unnecessary because we only needed num-min, num-max, and fold

blerner commented on src/web/arr/trove/chart.arr:
```diff
@@ -75,13 +86,150 @@ fun to-table3(xs :: List<Any>, ys :: List<Any>, zs :: List<Any>) -> TableIntern:
+    | is-stacked == 'absolute' then: 
+      # Find height of stack using sum functions
+      sum = {(l :: List<Number>): fold({(acc, elm): acc + elm}, 0, l)}
+      positive-only-sum = {(l :: List<Number>): sum(filter({(e): e >= 0}, l))}
+      negative-only-sum = {(l :: List<Number>): sum(filter({(e): e <= 0}, l))}
+      positive-sums = map(positive-only-sum, value-lists)
+      negative-sums = map(negative-only-sum, value-lists)
+      max-positive-height = num-max(0, get-with-cmp({(a, b): a > b}, positive-sums))
+      max-negative-height = num-min(0, get-with-cmp({(a, b): a < b}, negative-sums))
+      {max-positive-height; max-negative-height}
+
+    | otherwise: 
+      has-pos = any({(l): any( _ > 0, l)}, value-lists)
+      has-neg = any({(l): any( _ < 0, l)}, value-lists)
```
Be consistent in style: either ` > 0` or `{(e): e >= 0}`, but don't use both in such close proximity ;-)

JohnUUU replied:
That’s my bad, making it more consistent now

blerner commented on src/web/arr/trove/chart.arr:
```diff
@@ -122,6 +286,368 @@ y-max-method = method(self, y-max :: Number):
+format-axis-data-method = method(self, format-func :: (Number -> String)):
+  cases (Option) self.obj!axisdata: 
+    | none => 
+      raise("Should never have reached this point. Yell at John for not setting up the axis properties somewhere where he should have and please report this as a bug")
```
Clean up this error message...

JohnUUU replied:
Yeah really sorry about this, I meant to go back and change it, Doing that now!

JohnUUU committed:
Removed unnecessary helper functions and made code more concise + Cleaned up style and consistency

1. Removed Amaranth [The Color]
2. Replaced convert function with more succinct version
3. Fixed Unstable Reference to data parameter [data.if.length] by using a getter function instead
4. Replaced unnecessary get-with-cmp  helper function with fold
5. Minor edits to keep consistency with lambda style
6. Cleaned Up Error Message

JohnUUU committed:
Minor fix to error string

JohnUUU committed:
Fixed Color List + Pointers Combination

- There was an obscure bug where the combination of calling the ``colors`` method with a list larger than the number of “stacks/bars” in a multi-bar-chart series would interfere with calling the ```add-pointers``` method. Fixed this bug by slicing the color-list to match the number of “stacks/bars”.
- Also changed the behavior when the ``colors`` method is called on a list smaller than the number of “stacks/bars” in a multi-bar-chart series. Defaults to black for the colors not included in the color list now.

eyanje committed:
Replace unnecessary carets

JohnUUU committed:
Added Error Propagator in JS Chart Setup and Updated Pyret side type checking

1) Took the suggestion from Ben Lerner and copy and pasted the new setup code that would propagate the error from JS to Pyret.

2) Added type checking for all Bar chart methods that were Collection<Type> that failed to give meaningful error messages before.

JohnUUU committed:
Refined Types and Backup Error Catcher

1) From the meeting - We added a backup error catcher by propagating the error from JS and catching the error with run-task. Note this is still not super readable and everything should be caught in Pyret ideally. It's a backup thats just better than letting Pyret hang or log everything to console.

2) Type Refinement - We also added type refinements to each of the functions so that they could check the types of inputs wrapped in lists or options. Potential Problem: This still references the chart.arr file in the error and I can't seem to figure out how to get around it.

JohnUUU committed:
Used Refined Types in JS to type check instead

^ Created Refined Types for Lists of stuff or Options of Stuff in chart-lib.js,  and used them in Pyret to give slightly more concise error messages

eyanje committed:
Annotate and clean js errors (removes test code)

JohnUUU committed:
Changed default-color to color to keep it consistent with everything else

^ What the description said

JohnUUU committed:
Merge branch 'stacked-bar-chart' of https://github.com/JohnUUU/code.pyret.org into stacked-bar-chart

JohnUUU committed:
Added a .sort() method with default cmp of < and eq of ==

This is to add an easier to use, quick and dirty sort method for new students who might not want to deal with the idea of a cmp or eq.

JohnUUU committed:
Changed multi-bar-chart.colors behavior when less colors are given than number of components

1) Before when a multi-bar-chart series w/ n components called the **colors** method with m colors, where m < n, the rest of the components would all be colored black. This commit updates the code to default back to google chart default colors (Hard-Coded) instead!

JohnUUU committed:
Fixed Colors + Add-Pointers Integration

Turns out that the previous commit caused a change that disrupted the behavior when both colors and add-pointers was called together, This update fixes that disruption.

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -174,6 +214,116 @@
 
   //////////////////////////////////////////////////////////////////////////////
 
+  /**
```
It's probably worth saying up top that this was copied from a jsfiddle more explicitly.

JohnUUU replied:
The jsfiddle is actually from @eyanje so I'm not sure it's copied, probably more of an example. But it still might be worth putting up top for reference! Do you want me to do that in this PR or do you want to do it directly on horizon since this has already been merged. Sorry not 100% up to date on PR etiquette.

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -174,6 +214,116 @@
+    });
+  }
+
+  /**
```
For example, I think we wrote this function, but the block above came from somewhere else – that's useful to know when debugging later.

JohnUUU replied:
Yeah, but since the example is specific to the function, I think it actually makes sense to keep the jsfiddle comment above the function and maybe make a copy reference to it at the top if you want all the jsfiddles to be referenced at the top.

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -268,23 +418,259 @@
+      chartType: horizontal ? google.visualization.BarChart : google.visualization.ColumnChart,
       onExit: defaultImageReturn,
-      mutators: [axesNameMutator, yAxisRangeMutator],
+      mutatorgraphs: [axesNameMutator, yAxisRangeMutator],
```
Why is this key changed? Don't see `mutatorgraphs` mentioned elsewhere in the file.

JohnUUU replied:
I'm going to be 100% honest. I have no idea why, I think this might have been a change made in one of the meetings we had with @blerner since I don't recall really ever changing this on my own, and I don't really remember what it does. I would check in with ben!

sorawee replied:
@blerner can you check this? It looks incorrect to me.

Here's a program that demonstrates the bug:

```

include chart

a-stacked-bar-chart-series = from-list.stacked-bar-chart(
  [list: 'CA', 'TX', 'NY', 'FL', 'IL', 'PA'],
  [list:
    [list: 2704659,4499890,2159981,3853788,10604510,8819342,4114496],
    [list: 2027307,3277946,1420518,2454721,7017731,5656528,2472223],
    [list: 1208495,2141490,1058031,1999120,5355235,5120254,2607672],
    [list: 1140516,1938695,925060,1607297,4782119,4746856,3187797],
    [list: 894368,1558919,725973,1311479,3596343,3239173,1575308],
    [list: 737462,1345341,679201,1203944,3157759,3414001,1910571]],
  [list:
    'Under 5 Years',
    '5 to 13 Years',
    '14 to 17 Years',
    '18 to 24 Years',
    '25 to 44 Years',
    '45 to 64 Years',
    '65 Years and Over'])


render-chart(a-stacked-bar-chart-series)
  .x-axis("hello")
  .display()
```

blerner replied:
Joe and I saw that, and know it's wrong -- the question is whether it should be renamed back to `mutators`, or deleted entirely.  @JohnUUU needs to check it and see which behavior is intended.

JohnUUU replied:
Yeah definitely, I'll do that right now!

JohnUUU replied:
Oh I'm just realizing that new commits no longer get forwarded here since the PR has been merged. Should I open another PR? If you want to just make the change directly on horizon the right behavior happens when we rename it back to mutators.

blerner replied:
Yes, go ahead and make a new PR and add me and Joe as reviewers.

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -781,7 +1167,7 @@ ${labelRow}`;
 
       function setup(restarter) {
-        const tmp = f(globalOptions, rawData);
+        var tmp = f(globalOptions, rawData);
```
Is this an important `const` -> `var` change?

JohnUUU replied:
It doesn't seem to be, as far as I can tell the only things that we really edit are the variables of tmp which can still be changed if tmp is a const and tmp doesn't every really seem to be reassigned.

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -804,11 +1190,16 @@ ${labelRow}`;
+          try{
+            setup(restarter)
+          } catch (e) {
+            return restarter.error(e);
```
Thank you! It's so frustrating to debug when we miss errors like these.

JohnUUU replied:
No problem! It's all thanks to ben lerner's help!

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -827,17 +1218,24 @@ ${labelRow}`;
+      "LoS": ann("List<String>", checkListWith(RUNTIME.isString)), 
+      "LoN": ann("List<Number>", checkListWith(RUNTIME.isNumber)),
+      "LoLoN": ann("List<List<Number>>", checkListWith(checkListWith(RUNTIME.isNumber))),
+      "LoLoLoN": ann("List<List<List<<Number>>>", checkListWith(checkListWith(checkListWith(RUNTIME.isNumber)))),
```
Is the idea that the user will use these types?

JohnUUU replied:
The main idea behind this was that pyret doesn't typecheck nested types so many of these methods that would require List<Type> or Option<Type> Etc... would give fairly unhelpful type errors when the wrong nested type was given. This is just so that we can give better type errors in that edge case!

jpolitz commented:
I pushed my review and merged so we can play with this on live horizon, but I would like answers to the questions I posed! :-)
